change_detect.go

```diff
@@ -0,0 +1,139 @@
+package kapacitor
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/influxdata/kapacitor/edge"
+	"github.com/influxdata/kapacitor/keyvalue"
+	"github.com/influxdata/kapacitor/models"
+	"github.com/influxdata/kapacitor/pipeline"
+)
+
+type ChangeDetectNode struct {
+	node
+	d *pipeline.ChangeDetectNode
+}
+
+// Create a new changeDetect node.
+func newChangeDetectNode(et *ExecutingTask, n *pipeline.ChangeDetectNode, d NodeDiagnostic) (*ChangeDetectNode, error) {
+	dn := &ChangeDetectNode{
+		node: node{Node: n, et: et, diag: d},
+		d:    n,
+	}
+	// Create stateful expressions
+	dn.node.runF = dn.runChangeDetect
+	return dn, nil
+}
+
+func (n *ChangeDetectNode) runChangeDetect([]byte) error {
+	consumer := edge.NewGroupedConsumer(
+		n.ins[0],
+		n,
+	)
+	n.statMap.Set(statCardinalityGauge, consumer.CardinalityVar())
+	return consumer.Consume()
+}
+
+func (n *ChangeDetectNode) NewGroup(group edge.GroupInfo, first edge.PointMeta) (edge.Receiver, error) {
+	return edge.NewReceiverFromForwardReceiverWithStats(
+		n.outs,
+		edge.NewTimedForwardReceiver(n.timer, n.newGroup()),
+	), nil
+}
+
+func (n *ChangeDetectNode) newGroup() *changeDetectGroup {
+	return &changeDetectGroup{
+		n: n,
+	}
+}
+
+type changeDetectGroup struct {
+	n        *ChangeDetectNode
+	previous edge.FieldsTagsTimeGetter
+}
+
+func (g *changeDetectGroup) BeginBatch(begin edge.BeginBatchMessage) (edge.Message, error) {
+	if s := begin.SizeHint(); s > 0 {
+		begin = begin.ShallowCopy()
+		begin.SetSizeHint(s - 1)
+	}
```

Comment:
This size hinting is not relevant to the change detect node, meaning its now know that the batch size will be `s - 1`, but rather its unknown. As such the size hint should be set to 0 since an unknown number of points will be emitted.